metadata/src/main/java/org/apache/kafka/controller/ActivationRecordsGenerator.java

```diff
@@ -0,0 +1,197 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.controller;
+
+import org.apache.kafka.common.metadata.AbortTransactionRecord;
+import org.apache.kafka.common.metadata.BeginTransactionRecord;
+import org.apache.kafka.common.metadata.EndTransactionRecord;
+import org.apache.kafka.metadata.bootstrap.BootstrapMetadata;
+import org.apache.kafka.metadata.migration.ZkMigrationState;
+import org.apache.kafka.server.common.ApiMessageAndVersion;
+import org.apache.kafka.server.common.MetadataVersion;
+import org.slf4j.Logger;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class ActivationRecordsGenerator {
+
+    static ControllerResult<Void> recordsForEmptyLog(
+        Logger log,
+        long transactionStartOffset,
+        boolean zkMigrationEnabled,
+        BootstrapMetadata bootstrapMetadata,
+        MetadataVersion metadataVersion
+    ) {
+        List<ApiMessageAndVersion> records = new ArrayList<>();
+
+        if (transactionStartOffset != -1L) {
+            // In-flight bootstrap transaction
+            if (!metadataVersion.isMetadataTransactionSupported()) {
+                throw new RuntimeException("Detected partial bootstrap records transaction at " +
+                    transactionStartOffset + ", but the metadata.version " + metadataVersion +
+                    " does not support transactions. Cannot continue.");
+            } else {
+                log.warn("Detected partial bootstrap records transaction at " + transactionStartOffset +
+                    " during controller activation. Aborting this transaction and re-committing bootstrap records");
+                records.add(new ApiMessageAndVersion(
+                    new AbortTransactionRecord().setReason("Controller failover"), (short) 0));
+                records.add(new ApiMessageAndVersion(
+                    new BeginTransactionRecord().setName("Bootstrap records"), (short) 0));
+            }
+        } else if (metadataVersion.isMetadataTransactionSupported()) {
+            // No in-flight transaction
+            records.add(new ApiMessageAndVersion(
+                new BeginTransactionRecord().setName("Bootstrap records"), (short) 0));
+        }
+
+        // If no records have been replayed, we need to write out the bootstrap records.
+        // This will include the new metadata.version, as well as things like SCRAM
+        // initialization, etc.
+        log.info("The metadata log appears to be empty. Appending {} bootstrap record(s) " +
```

Comment:
I think it would be good to consolidate these log messages into one message that says:
- whether we had to abort a bootstrap transaction
- whether we're de-novo or migrating
- MV
- # of bootstrap records

and I guess that log message can be WARN. It is quite important, after all....